include/envoy/server/listener_manager.h

```diff
@@ -153,6 +153,15 @@ class ListenerManager {
    */
   virtual bool removeListener(const std::string& name) PURE;
 
+  /*
+   * Stops listeners from accepting new connections without actually removing them. This is
+   * used by /drain_listeners admin endpoint. This method directly stops the listeners on workers.
+   * Once a listener is stopped, any listener modifications are not allowed.
+   * @param inbound_only if true stops all the inbound listeners and stops all listeners if false.
+   * @return true if any listeners have been stopped.
+   */
+  virtual bool shutdownListeners(bool inbound_only) PURE;
```

Comment:
Can this API just be merged into `stopListeners()` below? Can that API take an enum of what type of stop to perform?

Reply:
We could do that and actually did think about it. But couple of points 
1. Do we have to maintain this api for any backward compatibility reasons because it is a public API?
2. When you mean stop state enum - are you thinking AllListeners, Inbound Listeners as values? I was thinking we could introduce a more generic “StopListenerSelector” that can have more wide categories of criteria like name etc instead of an enum - And if we want to do that, can we do that later? 
If you think more wide criteria is not needed at all, I can go with enum and change it now. Let me know

Reply:
And  no new changes are allowed to listeners after the `stopListeners` call - that would be behavioural change if we merge, but should be OK I guess?